programs/gummyroll/src/state/merkle_roll.rs

```diff
@@ -258,82 +326,114 @@ impl<const MAX_DEPTH: usize, const MAX_BUFFER_SIZE: usize> MerkleRoll<MAX_DEPTH,
-        msg!("Fast-forwarding proof, starting index {}", j);
+    ) -> bool {
         let mask: usize = MAX_BUFFER_SIZE - 1;
         let padding: usize = 32 - MAX_DEPTH;
+
+        let mut updated_leaf = leaf;
         sol_log_compute_units();
         // Modifies proof by iterating through the change log
         loop {
-            // If use_full_buffer is false, this loop will terminate if the initial value of j is the active index
-            if !use_full_buffer && j == self.active_index {
+            // If use_full_buffer is false, this loop will terminate if the initial value of changelog_buffer_index is the active index
+            if !use_full_buffer && changelog_buffer_index == self.active_index {
                 break;
             }
-            j += 1;
-            j &= mask as u64;
-            if index != self.change_logs[j as usize].index {
-                let common_path_len = ((index ^ self.change_logs[j as usize].index) << padding)
+            changelog_buffer_index += 1;
+            changelog_buffer_index &= mask as u64;
+            if leaf_index != self.change_logs[changelog_buffer_index as usize].index {
+                let common_path_len = ((leaf_index
+                    ^ self.change_logs[changelog_buffer_index as usize].index)
+                    << padding)
                     .leading_zeros() as usize;
                 let critbit_index = (MAX_DEPTH - 1) - common_path_len;
-                proof[critbit_index] = self.change_logs[j as usize].path[critbit_index];
+                proof[critbit_index] =
+                    self.change_logs[changelog_buffer_index as usize].path[critbit_index];
             } else {
-                updated_leaf = self.change_logs[j as usize].get_leaf();
+                updated_leaf = self.change_logs[changelog_buffer_index as usize].get_leaf();
             }
             // If use_full_buffer is true, this loop will do 1 full pass of the change logs
-            if use_full_buffer && j == self.active_index {
+            if use_full_buffer && changelog_buffer_index == self.active_index {
                 break;
             }
         }
         sol_log_compute_units();
-        let valid_root = recompute(updated_leaf, proof, index) == self.get_change_log().root;
-        if updated_leaf != leaf || index > self.rightmost_proof.index {
+        updated_leaf == leaf
+    }
+
+    /// Fast-forwards submitted proof to be valid for the root at `self.current_index`
+    ///
+    /// Updates proof & updates root & stores
+    ///
+    /// Takes in `j`, which is the root index that this proof was last valid for
+    fn update_and_apply_proof(
+        &mut self,
+        current_root: Node,
+        leaf: Node,
+        new_leaf: Node,
+        proof: &mut [Node; MAX_DEPTH],
+        leaf_index: u32,
+        append_on_conflict: bool,
+        use_full_buffer: bool,
+    ) -> Option<Node> {
+        let mask: usize = MAX_BUFFER_SIZE - 1;
+        let changelog_buffer_index: u64;
```

Comment:
```
let changelog_buffer_index = if use_full_buffer {
    ...
} else {
    ...
};
```

Reply:
good catch

Reply:
not that useful, now that we're throwing an error in the else block